File: hospital-ai/backend/app/schemas/research.py
"""Pydantic schemas for the Research Agent."""

from __future__ import annotations

from datetime import datetime
from typing import Any, Dict, List, Optional
from uuid import UUID

from pydantic import BaseModel, Field


class ResearchStartRequest(BaseModel):
    patient_id: UUID
    diagnosis_result_id: Optional[UUID] = None


class LiteratureItemOut(BaseModel):
    reference_id: str
    title: str
    authors: List[str] = Field(default_factory=list)
    journal: str = ""
    publication_year: int = 0
    study_type: str = ""
    summary: str = ""
    url: str = ""
    condition: str = ""
    relevance_score: float = 0.0


class ClinicalTrialItemOut(BaseModel):
    trial_id: str
    title: str
    status: str = ""
    phase: str = ""
    outcome_summary: str = ""
    eligibility_summary: str = ""
    condition: str = ""
    url: str = ""
    relevance_score: float = 0.0


class GuidelineItemOut(BaseModel):
    source: str
    title: str
    recommendation: str = ""
    condition: str = ""
    published_year: int = 0
    url: str = ""
    relevance_score: float = 0.0


class DrugEvidenceItemOut(BaseModel):
    drug_name: str
    condition: str = ""
    effectiveness_summary: str = ""
    known_side_effects: List[str] = Field(default_factory=list)
    contraindications: List[str] = Field(default_factory=list)
    drug_interactions: List[str] = Field(default_factory=list)
    supporting_evidence: str = ""
    relevance_score: float = 0.0


class ClinicalEvidenceOut(BaseModel):
    id: Optional[UUID] = None
    evidence_type: str
    condition: str
    title: str
    source: str
    reference_id: str = ""
    url: str = ""
    publication_date: Optional[str] = None
    summary: str = ""
    evidence_level: str = "Medium"
    relevance_score: float = 0.0
    confidence: float = 0.0


class ResearchRecommendationOut(BaseModel):
    condition: str
    supporting_literature: List[str] = Field(default_factory=list)
    clinical_guidelines: List[str] = Field(default_factory=list)
    evidence_summary: str = ""
    recommended_diagnostic_tests: List[str] = Field(default_factory=list)
    research_highlights: List[str] = Field(default_factory=list)
    confidence_score: float = 0.0


class ResearchResultOut(BaseModel):
    id: UUID
    patient_id: UUID
    diagnosis_result_id: Optional[UUID] = None
    conditions_researched_json: List[Any] = Field(default_factory=list)
    pubmed_json: List[Any] = Field(default_factory=list)
    clinical_trials_json: List[Any] = Field(default_factory=list)
    guidelines_json: List[Any] = Field(default_factory=list)
    drug_efficacy_json: List[Any] = Field(default_factory=list)
    evidence_ranking_summary_json: Dict[str, Any] = Field(default_factory=dict)
    recommendation_json: List[Any] = Field(default_factory=list)
    summary: Optional[str] = None
    provider: str = "mock"
    status: str = "Completed"
    error_message: Optional[str] = None
    processing_time_ms: Optional[int] = None
    created_at: datetime
    updated_at: Optional[datetime] = None


class ResearchStartResponse(BaseModel):
    patient_id: UUID
    diagnosis_result_id: Optional[UUID] = None
    status: str = "Completed"
    processing_time_ms: int
    summary: str
    provider: str
    warnings: List[str] = Field(default_factory=list)
    conditions_researched: List[str] = Field(default_factory=list)
    pubmed_results: List[LiteratureItemOut] = Field(default_factory=list)
    clinical_trials: List[ClinicalTrialItemOut] = Field(default_factory=list)
    guidelines: List[GuidelineItemOut] = Field(default_factory=list)
    drug_efficacy: List[DrugEvidenceItemOut] = Field(default_factory=list)
    evidence: List[ClinicalEvidenceOut] = Field(default_factory=list)
    evidence_level_counts: Dict[str, int] = Field(default_factory=dict)
    recommendations: List[ResearchRecommendationOut] = Field(default_factory=list)
    research_result: ResearchResultOut


class ResearchHistoryItemOut(BaseModel):
    id: UUID
    created_at: datetime
    summary: Optional[str] = None
    conditions_researched: List[str] = Field(default_factory=list)
    status: str = "Completed"

File: hospital-ai/frontend/src/types/research.ts
/** Research Agent types — evidence enrichment for Diagnosis Agent results. */
import type { OrchestratorDebugInfo } from '@/types/aiOrchestrator';

export type EvidenceLevel = 'High' | 'Medium' | 'Low';

export const EVIDENCE_LEVEL_TONE: Record<
  string,
  'green' | 'blue' | 'amber' | 'red' | 'gray'
> = {
  High: 'green',
  Medium: 'amber',
  Low: 'gray',
};

export interface LiteratureItem {
  reference_id: string;
  title: string;
  authors: string[];
  journal: string;
  publication_year: number;
  study_type: string;
  summary: string;
  url: string;
  condition: string;
  relevance_score: number;
}

export interface ClinicalTrialItem {
  trial_id: string;
  title: string;
  status: string;
  phase: string;
  outcome_summary: string;
  eligibility_summary: string;
  condition: string;
  url: string;
  relevance_score: number;
}

export interface GuidelineItem {
  source: string;
  title: string;
  recommendation: string;
  condition: string;
  published_year: number;
  url: string;
  relevance_score: number;
}

export interface DrugEvidenceItem {
  drug_name: string;
  condition: string;
  effectiveness_summary: string;
  known_side_effects: string[];
  contraindications: string[];
  drug_interactions: string[];
  supporting_evidence: string;
  relevance_score: number;
}

export interface ClinicalEvidence {
  id?: string;
  evidence_type: 'pubmed' | 'clinical_trial' | 'guideline' | 'drug_efficacy';
  condition: string;
  title: string;
  source: string;
  reference_id: string;
  url: string;
  publication_date?: string | null;
  summary: string;
  evidence_level: EvidenceLevel | string;
  relevance_score: number;
  confidence: number;
}

export interface ResearchRecommendation {
  condition: string;
  supporting_literature: string[];
  clinical_guidelines: string[];
  evidence_summary: string;
  recommended_diagnostic_tests: string[];
  research_highlights: string[];
  confidence_score: number;
}

export interface ResearchResult {
  id: string;
  patient_id: string;
  diagnosis_result_id?: string | null;
  conditions_researched_json: string[];
  pubmed_json: LiteratureItem[];
  clinical_trials_json: ClinicalTrialItem[];
  guidelines_json: GuidelineItem[];
  drug_efficacy_json: DrugEvidenceItem[];
  evidence_ranking_summary_json: Record<string, number>;
  recommendation_json: ResearchRecommendation[];
  summary?: string | null;
  provider: string;
  status: string;
  error_message?: string | null;
  processing_time_ms?: number | null;
  created_at: string;
  updated_at?: string | null;
}

export interface ResearchStartResult {
  patient_id: string;
  diagnosis_result_id?: string | null;
  status: string;
  processing_time_ms: number;
  summary: string;
  provider: string;
  warnings: string[];
  conditions_researched: string[];
  pubmed_results: LiteratureItem[];
  clinical_trials: ClinicalTrialItem[];
  guidelines: GuidelineItem[];
  drug_efficacy: DrugEvidenceItem[];
  evidence: ClinicalEvidence[];
  evidence_level_counts: Record<string, number>;
  recommendations: ResearchRecommendation[];
  research_result: ResearchResult;
  ai_debug: OrchestratorDebugInfo[];
}

export interface ResearchHistoryItem {
  id: string;
  created_at: string;
  summary?: string | null;
  conditions_researched: string[];
  status: string;
}
